applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -0,0 +1,269 @@
+
+static const char *get_dvfs_frequency_setting_name(enum dvfs_frequency_setting setting)
+{
+	switch (setting) {
+	case DVFS_FREQ_HIGH: return "DVFS_FREQ_HIGH";
+	case DVFS_FREQ_MEDLOW: return "DVFS_FREQ_MEDLOW";
+	case DVFS_FREQ_LOW: return "DVFS_FREQ_LOW";
+	default: return "Unknown";
+	}
+}
+
+static const char *get_dvfs_state_name(enum dvfs_state state)
+{
+	switch (state) {
+	case DVFS_STATE_INITIALIZING: return "DVFS_STATE_INITIALIZING";
+	case DVFS_STATE_LLPM_CONNECTED: return "DVFS_STATE_LLPM_CONNECTED";
+	case DVFS_STATE_USB_CONNECTED: return "DVFS_STATE_USB_CONNECTED";
+	default: return "Unknown";
+	}
+}
+
+static void cancel_dvfs_retry_work(void)
+{
+	(void) k_work_cancel_delayable(&dvfs_retry_data.dvfs_retry_work);
+	dvfs_retry_data.dfvs_retries_cnt = 0;
+}
+
+static void handle_dvfs_error(int32_t err)
+{
+	if (dvfs_retry_data.dfvs_retries_cnt > DVFS_NUMBER_OF_RETRIES) {
+		LOG_ERR("DVFS retry count exceeded.");
+		module_set_state(MODULE_STATE_ERROR);
+		module_state = STATE_ERROR;
+		cancel_dvfs_retry_work();
+		return;
+	}
+	dvfs_retry_data.dfvs_retries_cnt++;
+	k_timeout_t timeout;
+
+	if (err == -EBUSY) {
+		LOG_DBG("DVFS frequency change in progress");
+		timeout = DVFS_RETRY_BUSY_TIMEOUT;
+	} else if (err == -EAGAIN) {
+		LOG_DBG("DVFS not initialized, trying again.");
+		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
+	} else {
+		LOG_ERR("DVFS freq change returned with error: %d", err);
+		module_set_state(MODULE_STATE_ERROR);
+		module_state = STATE_ERROR;
+		cancel_dvfs_retry_work();
+		return;
+	}
+	(void) k_work_reschedule(&dvfs_retry_data.dvfs_retry_work, timeout);
+}
+
+static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+{
+	int32_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+
+	if (ret) {
+		handle_dvfs_error(ret);
+	} else {
+		current_freq = target_freq;
+		LOG_INF("Have requested %s frequency",
+			get_dvfs_frequency_setting_name(target_freq));
+		cancel_dvfs_retry_work();
+	}
+}
+
+static enum dvfs_frequency_setting check_required_frequency(void)
+{
+	if (dfvs_requests_state & dvfs_high_freq_bitmask) {
+		return DVFS_FREQ_HIGH;
+	} else if (dfvs_requests_state & dvfs_medlow_freq_bitmask) {
+		return DVFS_FREQ_MEDLOW;
+	} else {
+		return DVFS_FREQ_LOW;
+	}
+}
+
+static void process_dvfs_states(enum dvfs_state state, bool turn_on)
+{
+	if (module_state == STATE_ERROR) {
+		return;
+	}
+
+	if (turn_on) {
+		dfvs_requests_state |= BIT(state);
+		LOG_DBG("%s ACTIVE", get_dvfs_state_name(state));
+	} else {
+		dfvs_requests_state &= ~BIT(state);
+		LOG_DBG("%s NOT ACTIVE", get_dvfs_state_name(state));
+	}
+
+	enum dvfs_frequency_setting required_freq = check_required_frequency();
+
+	if ((required_freq != current_freq) &&
+	    (!k_work_delayable_is_pending(&dvfs_retry_data.dvfs_retry_work))) {
+		set_dvfs_freq(required_freq);
+	} else if ((required_freq == current_freq) &&
+	    (k_work_delayable_is_pending(&dvfs_retry_data.dvfs_retry_work))) {
+		cancel_dvfs_retry_work();
+	}
```

Comment:
During the standup, you discussed some performance concerns related to the module. Consider removing the `else` part and handle this in the work handler (zero retry counter + early return) to avoid impacting performance (the `else` case may be entered quite often by the code after we introduce handling of events that appear often - e.g. config channel or SMP DFU).

Reply:
I do not expect it to happen often. And if it doesn't happen often it actually boost performance because we won't enter work handler at all. But we might see how it behaves when we introduce new functionality.

Reply:
We can leave it as is for now.